base/src/commonMain/kotlin/state/State.kt

```diff
@@ -234,31 +241,113 @@ abstract class StateRepo<S : State>(coroutineContext: CoroutineContext = Dispatc
+inline fun <S> defaultLazySharedFlow():Lazy<MutableSharedFlow<S>> = lazy { MutableSharedFlow(replay = 1, onBufferOverflow = BufferOverflow.DROP_OLDEST) }
+
 /**
  * A [StateRepo] that represents its [State] as a Hot flow.
  */
-abstract class HotStateRepo<S : State>(coroutineContext: CoroutineContext = Dispatchers.Main) : StateRepo<S>(coroutineContext) {
+abstract class HotStateRepo<S : State>(
+    coroutineContext: CoroutineContext = Dispatchers.Main.immediate,
+    private val lazyMutableSharedFlow: Lazy<MutableSharedFlow<S>> = defaultLazySharedFlow()
+) : StateRepo<S, MutableSharedFlow<S>>(coroutineContext) {
+
+    // guards once only initialization across threads
+    private val initialized = AtomicBoolean(false)
+
+    override val mutableFlow: MutableSharedFlow<S>
+    get() {
+        val isInitialized = lazyMutableSharedFlow.isInitialized()
+        val flow = lazyMutableSharedFlow.value
+        if (!isInitialized && initialized.compareAndSet(expected = false, new = true))
+            launch(coroutineContext) {
+                mutableFlow.subscriptionCount.filter { it < 1 }.first()
+                initialize()
+            }
+        return flow
+    }
+}
+
+abstract class BaseColdStateRepo<S:State, F:MutableSharedFlow<S>>(
+    context: CoroutineContext = Dispatchers.Main.immediate
+) : StateRepo<S, F>(context) {
+
+    abstract val lazyMutableFlow: Lazy<F>
 
-    private val hotFlowable = lazy {
-        HotFlowable(runBlocking { initialize() })
+    override val mutableFlow: F by lazy {
+        val flow by lazyMutableFlow
+        launch(coroutineContext) {
+            flow.onCollectionEvent {
+                when (it) {
+                    NoMoreCollections -> deinitialize(state())
+                    FirstCollection -> firstInitialization()
+                    LaterCollections -> laterInitializations()
+                }
+            }
+        }
+        flow
     }
-    override val flowable: BaseFlowable<S> get() = hotFlowable.value
+
+    abstract suspend fun firstInitialization()
+
+    abstract suspend fun laterInitializations()
+
+    abstract suspend fun deinitialize(state: S)
 }
 
+open class ColdStateFlowRepo<S:State>(
+    coroutineContext: CoroutineContext = Dispatchers.Main.immediate,
+    val init: suspend () -> S,
+    val deinit: suspend (S) -> Unit
+) : BaseColdStateRepo<S, MutableStateFlow<S>>(
+    context = coroutineContext,
+) {
+    val stateflow: StateFlow<S>
+        get() = mutableFlow
+
+    // the first initialization is done in the lazy block below since StateFlow requires an initial value
+    override suspend fun initialValue(): S = init()
+
+    override suspend fun deinitialize(state: S) {
+        deinit(state)
+    }
+
+    override val lazyMutableFlow: Lazy<MutableStateFlow<S>> =
+        lazy {
+            runBlocking(this.coroutineContext) {
+                MutableStateFlow(
+                    init()
```

Comment:
`initialValue()`?

Reply:
no, `init` is a closure passed from the constructor, this flexibility was needed to support the legacy API in a class that extends this one.